### DISCUSSION TITLE
Doctrine `getReference` method not working

### DISCUSSION THREAD

**Discussion by bogdandubyk** [symfony/symfony#42822] [Q&A]
Depends on https://www.doctrine-project.org/projects/doctrine-orm/en/2.9/reference/advanced-configuration.html#reference-proxies in case if you want to add some association without actually loading the entity from the database, you can use `getReference` method.

So in one of my classes to which I'm passing `EntityManagerInterface` as a dependency I'm calling `$this->entityManager->getReference(Entity::class, $id);`, but it's failing with exception:

```
#message: "Undefined method "getReference". The method name must start with either findBy, findOneBy or countBy!"
  #code: 0
  #file: "/srv/catalog/app/vendor/doctrine/orm/lib/Doctrine/ORM/EntityRepository.php"
```

it's because `Doctrine\ORM\EntityRepository` doesn't have `getReference` method, but I do not understand why the application using `Doctrine\ORM\EntityRepository` when I'm injecting `EntityManagerInterface`??? As I can see `Doctrine\ORM\EntityRepository` even not implementing `EntityManagerInterface`???

How can I fix that, or maybe in Symfony it makes no sense to use `getReference` as it's by default returning reference until you access some data on that reference (lazy loading), but looks like that's not true because dumping doctrine SQL logger I can see query to get data from that table

**Comment by stof** [comment]
Can you share the failing code ?

**Reply by bogdandubyk** [reply]
```
class SomeClass
{
    public function __construct(
        private EntityManagerInterface $entityManager
    ) {
    }

    public function someMethod(EntityToUpdate $baseEntity, int $id): void 
    {
         ...
          $entity = $this->entityManager->getReference(Entity::class, $id); <--- failing here
          $baseEntity->changeEntity($entity);
          $this->entityManager->persist($baseEntity);
          $this->entityManager->flush()
         ... 
    }
```

**Reply by stof** [reply]
This looks weird to me. The property is typed, so it is impossible to get a repository in it. Are you sure that this `getReference` call is the failing one (look at the next parts of the stack trace, that you truncated in your question) ?

Also, try dumping `get_class($this->entityManager)` to see whether it is the `Doctrine\ORM\EntityManager` (or the proxy generated for it by DoctrineBundle to allow resetting it) or whether you have a decorator applied on it that might do wrong things.

**Reply by bogdandubyk** [reply]
I tried `get_class($this->entityManager)` and yeah it's an instance of `Doctrine\ORM\EntityManager`, let me check a trace, but anyway, the exception message is `"Undefined method "getReference". The method name must start with either findBy, findOneBy or countBy!"` so looks like it is the issue...

**Reply by stof** [reply]
well, the error is happening in `Doctrine\ORM\EntityRepository`, while you confirmed that this code calls it on `Doctrine\ORM\EntityManager`. So that's making me a lot more suspicious about the fact that the undefined method call corresponds to this line of code.

**Reply by bogdandubyk** [reply]
I do not know what to tell... it's working... I already spend few hours testing it and trying to reproduce and it's working now.... I can show you the original exception with a trace, and it shows to that line of code.. but it seems the issue has gone without changing any line of code...

**Comment by stof** [comment]
Note that using `getReference` in Doctrine is most of the time a mistake. The prerequisite for using it is being **100% sure** that the id corresponds to a row in the DB (allowing Doctrine to skip querying the DB to verify that). If that prerequisite is not satisfied, the code will break in very weird ways. When you are not able to be sure about that, use the `find()` method (for which Doctrine already optimises things if it already loaded the object with that id)

**Reply by bogdandubyk** [reply]
I'm sure it exists, at least in this specific case, but thank you I'll make a note and rethink if I need `getReference`. But still trying to understand why its' not working

**Reply by aszenz** [reply]
> Note that using `getReference` in Doctrine is most of the time a mistake. The prerequisite for using it is being **100% sure** that the id corresponds to a row in the DB (allowing Doctrine to skip querying the DB to verify that). If that prerequisite is not satisfied, the code will break in very weird ways. When you are not able to be sure about that, use the `find()` method (for which Doctrine already optimises things if it already loaded the object with that id)

Is that also the case when u have foreign keys for id, we use getReference a lot, since we have foreign keys which enforces data consistency

**Reply by stof** [reply]
@aszenz if you map the foreign keys as Doctrine associations, the ORM itself will use `getReference` (or actually the lower-level infrastructure powering it). The only case for which I see `getReference` being valid is when you have a foreign key (to give you the necessary certainty) without using the Doctrine association mapping. But I don't see any use case for that, as it is basically reimplementing the `@ManyToOne` (or `@OneToOne`) of Doctrine yourselves instead of using the built-in feature.

**Reply by aszenz** [reply]
> @aszenz if you map the foreign keys as Doctrine associations, the ORM itself will use `getReference` (or actually the lower-level infrastructure powering it). The only case for which I see `getReference` being valid is when you have a foreign key (to give you the necessary certainty) without using the Doctrine association mapping. But I don't see any use case for that, as it is basically reimplementing the `@ManyToOne` (or `@OneToOne`) of Doctrine yourselves instead of using the built-in feature.

We have the foreign keys mapped as associations, say we get a currency id back from the api how do i set the invoice entities association to this currency id without having to fetch the whole currency entity. I'm primarily using getReference to set such associations.
Maybe I'm missing something here or there's a better way.
